### PR TITLE
Pre-declare all core dispatch table functions, and fix the internal ones

### PR DESCRIPTION
**Pull request by levitte**
When assigning pointers to functions in an OSSL_DISPATCH table, we try
to ensure that those functions are properly defined or declared with
an extra declaration using the corresponding function typedefs that
are defined by include/openssl/core_dispatch.h.

For the core dispatch table, found in crypto/provider_core.c, it seems
we forgot this habit, and thus didn't ensure well enough that the
function pointers that are assigned in the table can actually be used
for those dispatch table indexes.

This change adds all the missing declarations, and compensates for
differences with functions that do the necessary casting, making those
explicit rather than implicit, thereby trying to assure that we know
what we're doing.

One function is not fixed in this change, because there's a controversy,
a clash between the signature of BIO_ctrl() and OSSL_FUNC_BIO_ctrl_fn.
They have different return types.

### CODE REVIEW

**Comment by levitte** [comment]
Regarding the `BIO_ctrl` vs `OSSL_FUNC_BIO_ctrl_fn` clash, the easy
(and most correct) fix would be this:

``` diff
diff --git a/include/openssl/core_dispatch.h b/include/openssl/core_dispatch.h
index 6aeb9e61e1..0cf9f94dd6 100644
--- a/include/openssl/core_dispatch.h
+++ b/include/openssl/core_dispatch.h
@@ -173,8 +173,8 @@ OSSL_CORE_MAKE_FUNC(int, BIO_vprintf, (OSSL_CORE_BIO *bio, const char *format,
                                        va_list args))
 OSSL_CORE_MAKE_FUNC(int, BIO_vsnprintf,
                    (char *buf, size_t n, const char *fmt, va_list args))
-OSSL_CORE_MAKE_FUNC(int, BIO_ctrl, (OSSL_CORE_BIO *bio,
-                                    int cmd, long num, void *ptr))
+OSSL_CORE_MAKE_FUNC(long, BIO_ctrl, (OSSL_CORE_BIO *bio,
+                                     int cmd, long num, void *ptr))
 
 #define OSSL_FUNC_SELF_TEST_CB               100
 OSSL_CORE_MAKE_FUNC(void, self_test_cb, (OPENSSL_CORE_CTX *ctx, OSSL_CALLBACK **cb,
```

However, since this is a public facing interface, this change is
problematic.  Ideas welcome.

**Comment by levitte** [comment]
The `BIO_ctrl()` vs `OSSL_FUNC_BIO_ctrl_fn` clash is the only reason this is a draft.

**Comment by levitte** [comment]
Still looking for ideas re `BIO_ctrl`

**Comment by levitte** [comment]
> Still looking for ideas re `BIO_ctrl`

I gave that some thought, and added a commit for it.  The solution may be controversial

**Comment by levitte** [comment]
> Except I'd do s/BOGUS/INT/

That would suggest that it has proper usage, just happens to be "the int version".  I want it to be understood that *it shouldn't be used*...

File: include/openssl/core_dispatch.h
Comment (mattcaswell) [inline include/openssl/core_dispatch.h:179]: This is still a breaking change, e.g. consider the following:

```` C
#include <openssl/core.h>
#include <openssl/core_dispatch.h>

static int (*c_bio_ctrl)(OSSL_CORE_BIO *bio, int cmd, long num, void *ptr) = NULL;

int OSSL_provider_init(const OSSL_CORE_HANDLE *handle, const OSSL_DISPATCH *in,
                       const OSSL_DISPATCH **out, void **provctx)
{
    for (; in->function_id != 0; in++) {
        if (in->function_id == OSSL_FUNC_BIO_CTRL)
            c_bio_ctrl = OSSL_FUNC_BIO_ctrl(in);
    }

    /* Do stuff */

    return 1;
}
````
Compiles just fine before this change. Afterwards we get:

````
ctrl.c: In function ‘OSSL_provider_init’:
ctrl.c:20:24: warning: assignment to ‘int (*)(OSSL_CORE_BIO *, int,  long int,  void *)’ {aka ‘int (*)(struct ossl_core_bio_st *, int,  long int,  void *)’} from incompatible pointer type ‘long int (*)(OSSL_CORE_BIO *, int,  long int,  void *)’ {aka ‘long int (*)(struct ossl_core_bio_st *, int,  long int,  void *)’} [-Wincompatible-pointer-types]
   20 |             c_bio_ctrl = OSSL_FUNC_BIO_ctrl(in);
      |                        ^
````

**Comment by levitte** [comment]
I added a squash commit that makes for a leaner resolution of the BIO_ctrl() issue

**Review by mattcaswell — DISMISSED** [review]

**Comment by levitte** [comment]
@openssl/committers, second review desired

**Comment by mattcaswell** [comment]
OTC: @levitte to create a new PR to make the alternatives visible

**Comment by levitte** [comment]
This is now self contained and doesn't touch BIO_ctrl at all.  #18583 and #18584 are the two solutions that do

**Comment by t8m** [comment]
@mattcaswell please re-approve

**Comment by levitte** [comment]
Or in fact, any @openssl/otc member could do the second approval

**Comment by t8m** [comment]
Even committer could approve.

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Merged

master:
9574842e90e29015daa2b071e965cec9aa885c17 Pre-declare all core dispatch table functions, and fix the internal ones

3.0:
372f4b80e35954d19ffb255692bdd6e4807e5f7a Pre-declare all core dispatch table functions, and fix the internal ones

**Pull request closed without merging**
